File: CodeTest/src/ReetCode/Easy/1101to1200/Q1189_MaximumNumberOfBalloons.java
/**
* Author    : Lee In Bok 
* Date      : 2023.11.26(Sun)
* Spend Time: 07m 25s
* Runtime   : 13 ms (Beats 9.34%)
* Memory    : 42.1 MB (Beats 7.12%)
* Algoritm  : Hashing
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Q1189_MaximumNumberOfBalloons {
    public static void main(String[] args) {
        Solution s = new Solution();
        int result = s.maxNumberOfBalloons("loonbalxballpoon");

        System.out.println(result);
    }  
}

class Solution {
    public int maxNumberOfBalloons(String text) {
        Map<Character, Integer> map = new HashMap<>();
        List<Integer> list = new ArrayList<>();

        for(int i = 0; i < text.length(); i++){
            char key = text.charAt(i);

            map.put(key, map.getOrDefault(key, 0) + 1);
        }

        list.add(map.getOrDefault('b', 0));
        list.add(map.getOrDefault('a', 0));
        list.add(map.getOrDefault('l', 0) / 2);
        list.add(map.getOrDefault('o', 0) / 2);
        list.add(map.getOrDefault('n', 0));

        return list.stream().mapToInt(Integer::intValue).min().getAsInt();
    }
}

File: CodeTest/src/ReetCode/Easy/1101to1200/Q1189/Q1189_Maximum Number of Balloons.py
"""
# Author    : Lee In Bok 
# Date      : 2023.11.26(Sun)
# Spend Time: 07m 25s
# Runtime   : 35 ms (Beats 88.6%)
# Memory    : 16.4 MB (Beats 6.62%)
# Algoritm  : Dictionary
"""

from collections import Counter

class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        ans = 0
        target = ['b', 'a', 'l', 'l', 'o', 'o', 'n']
        dic = Counter(text)
        
        while True:
            for ch in target:
                if ch in dic and dic[ch] != 0:
                    dic[ch] -= 1
                else:
                    return ans
            ans += 1

s = Solution()
result = s.maxNumberOfBalloons("loonbalxballpoon")

print(result)